content/blog/2020-11-16-reproducibility-in-ml-my-tools/index.en.md

```diff
@@ -0,0 +1,122 @@
+[MLflow](https://mlflow.org/docs/latest/index.html) is an open-source platform for tracking experiments. It helps record and compare both the parameters and the results associated to different experiments. Using MLflow, you will get a visual interface (GUI) (Figure 1) to compare your experiments based on their parameters or their results.
+
+![](mlflow-ui.png)
+Figure 1: [A snapshot of the visual interface of MLflow](https://databricks.com/blog/2018/06/05/introducing-mlflow-an-open-source-machine-learning-platform.html).
+
+MLflow is the solution I use right now. It is minimal, easy to use and needs relatively low code (a couple of lines depending on the training framework you are using). 
+
+It can also do other things such as deploying and providing a central storage space for models, but those solutions are priced.
+
+### Weights & Biases
+[Weights & Biases (W&B)](https://docs.wandb.ai/) is an open-source platform to track machine learning experiments, visualize metrics and share results. Using W&B, you will get the same features as those presented for MLflow (see Figure 2 for a snapshot of the GUI). The difference is that the dashboard is more advanced for comparing metrics and logging artifacts such as specific data point predictions or models.
+
+![](wnb.png)
+Figure 2: [A snapshot of the visual interface of W&B](https://github.com/wandb/client#try-in-a-colab-).
+
+That being said, the two solutions offer great features, and there are plenty of other interesting solutions out there. I think you should use any of these solutions to manage your results and experiments as long as it suits your need.
+
+## Managing Configurations
+Managing my configurations was the problem I've faced the most among all my work. Machine learning models have plenty of parameters, and their settings can greatly impact performance. You might want to try different architectures, optimizers, and do a grid search on parameters. The problem is that it is difficult to easily swap settings between and adding new one can be cumbersome. My first solutions were [Argparse](https://docs.python.org/3/howto/argparse.html) and [Configparser](https://docs.python.org/3/library/configparser.html). These solutions are really useful since they are easy to set up, and arguments are readable. But the more you add parameters, the more the code becomes complex, and it becomes difficult to find yourself. Also, it feels strange to always just copypaste code to create new arguments. It feels contradictory to the [Rules of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)#:~:text=It%20states%20that%20two%20instances,and%20attributed%20to%20Don%20Roberts) idea that "If you copy-paste some bloc of code three times, you might create a function." I also tried other solutions such as [Sacred](https://pypi.org/project/sacred/) with JSON configuration files, but the framework is not well design in my opinion. In summary, something is missing in all of these solutions. 
```

Comment:
```suggestion
Managing my configurations was the problem I've faced the most during my work. Machine learning models have plenty of parameters, and their settings can greatly impact performance. You might want to try different architectures, optimizers, and do a grid search on parameters. The problem is that it is difficult to swap between settings and adding new one can be cumbersome. My first solutions were [Argparse](https://docs.python.org/3/howto/argparse.html) and [Configparser](https://docs.python.org/3/library/configparser.html). These solutions are really useful since they are easy to set up, and arguments are readable. But the more you add parameters, the more the code becomes complex, and it becomes difficult to navigate it. Also, it feels strange to always just copypaste code to create new arguments, somewhat contradictory to the [Rules of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)#:~:text=It%20states%20that%20two%20instances,and%20attributed%20to%20Don%20Roberts) idea that "If you copy-paste some bloc of code three times, you might create a function." I also tried other solutions such as [Sacred](https://pypi.org/project/sacred/) with JSON configuration files, but the framework is not well designed in my opinion. In summary, something is missing in all of these solutions. 
```

Reply:
Sorry c'est pas évident les changements. Les phrases changées sont:
- Managing my conf...
- The problem is that it is....
- But the more you add...
- Also, it feels...
- I also tried....